T&A: Revision and Enhancement of the ‘Personal Default Settings‘ -> Personal Test Settings Templates

lukas-heinrich opened the pull request:
Hi everyone,

This PR relates to the restructuring of ~~Personal Default Settings~~ _Personal Test Settings Templates_, as described in the feature Wiki [here](https://docu.ilias.de/go/wiki/wpage_8640_1357). These changes include the following steps:

- Moving the database structure for test settings to a new table and adapting the associated DTOs and repositories
- Migrating the personal test settings templates from the PHP-*serialized* structure into the new table schema
- Integrating the legacy table into the new UI `DataTable`
- Moving from legacy forms and modals to the new UI components
- Implementing an import/export functionality for templates

The changes in this PR have been tested internally and approved through a review by @thojou . 
As always, I look forward to feedback on this proposal. Thanks in advance!

Best
@lukas-heinrich

lukas-heinrich committed:
refactor: remove the dependency on a specific test from the test settings to improve the coupling of the DTOs

lukas-heinrich committed:
refactor: adjust settings repositories

lukas-heinrich committed:
chore: introduce database migration

- A new database table `tst_test_settings` will be introduced
- A reference ID to the corresponding settings will be stored in the `tst_tests` table
- A reference ID to these settings will be stored in the `tst_test_defaults` table
- Settings currently stored in `tst_tests` will be migrated to the new `tst_test_settings` table
- Default settings from `tst_test_defaults` will also be migrated to `tst_test_settings`

lukas-heinrich committed:
refactor: remove legacy datetime storage format

lukas-heinrich committed:
refactor: introduce PersonalSettingsTemplateRepository

lukas-heinrich committed:
feat: introduce personal settings template table and adjust naming

lukas-heinrich committed:
refactor: introduce factories to streamline data object creation

lukas-heinrich committed:
feat: introduce show template & delete template table actions

lukas-heinrich committed:
feat: introduce normalization to enhance export/import

- Introduced `PropertyNormalizer` trait for automatic property normalization.
- Added `Normalizable` and `ContextualNormalizable` interfaces to define normalization contracts.
- Implemented normalization methods in `Mark`, `MarkSchema`, `Settings`, and other related classes to facilitate data transformation.
- Enhanced `ScoreSettings`, `SettingsGamification`, and `SettingsResultDetails` with normalization capabilities.
- Updated `PersonalSettingsTemplate` to utilize the new normalization approach.

lukas-heinrich committed:
feat: add PersonalSettingsTableApplyAction and fix naming in PersonalSettingsTable

- Introduced `PersonalSettingsTableApplyAction` to handle template application actions.
- Updated the label in `PersonalSettingsTable` from 'tst_defaults_available' to 'personal_settings_templates_available'.
- Removed obsolete `ilTestPersonalDefaultSettingsTableGUITest` file.

lukas-heinrich committed:
feat: prepare personal settings template export

- Added `PersonalSettingsExporter` class to handle the export of personal settings templates to XML format.
- Introduced `getTemplateById` method in `PersonalSettingsRepository` to retrieve templates by their ID.
- Updated normalization logic in `MarkSchema` to handle missing `test_id` gracefully.

lukas-heinrich committed:
feat: introduce personal settings export functionality

- Added `PersonalSettingsTableExportAction` class to facilitate the export of personal settings templates.
- Updated `PersonalSettingsExporter` to include template ID handling and improved XML delivery logic.
- Integrated new exporter into the dependency injection container for streamlined access.

lukas-heinrich committed:
feat: introduce personal settings importer

- Added `PersonalSettingsImporter` class to facilitate the import of personal settings templates from XML.
- Updated `PersonalSettingsRepository` to support importing templates and managing main and score settings.
- Enhanced `PersonalSettingsExporter` to include ILIAS version in the exported XML.
- Introduced XML schema definition for personal settings templates to ensure validation during import.

lukas-heinrich committed:
refactor: streamline settings management and fix foreign key restrictions

- Deleted the `ilTestPersonalDefaultSettingsTableGUI` class as part of the refactoring process.
- Updated `PersonalSettingsRepository` to utilize the new `SettingsFactory` for creating templates.
- Enhanced template handling in `PersonalSettingsImporter` and `PersonalSettingsRepository` to streamline the import and management of personal settings.
- Adjusted database update steps to change foreign key constraints from RESTRICT to CASCADE for better integrity during deletions.

lukas-heinrich committed:
refactor: finalize personal settings table actions

- Added new repositories for personal settings and main settings to ilObjTestGUI.
- Implemented methods to show and apply personal settings templates.
- Refactored template handling in defaultsObject and added new actions for template management.
- Updated TabsManager to streamline access control for personal settings actions.
- Improved error handling in PersonalSettingsTableActions for better user feedback.

lukas-heinrich committed:
feat: add missing arguments and table columns

- Updated `createTemplateFor` method in `PersonalSettingsRepository` to include `description` and `author` parameters for better template metadata.
- Modified database update steps to add new `description` and `author` columns in the `tst_test_defaults` table, improving the template's data structure.

lukas-heinrich committed:
feat: introduce CreateTemplate-action and fix language keys

- Added `createTemplateCmd` method to `ilObjTestGUI` for creating new personal settings templates.
- Updated command handling to include new actions for showing and creating templates.
- Refactored `showPersonalSettingsCmd` to `showTemplatesCmd` for better clarity.
- Introduced `PersonalSettingsCreateAction` class to manage the creation of personal settings templates.
- Updated language files to reflect new terminology and messages related to personal settings.

lukas-heinrich committed:
refactor: integrate changes into test module

lukas-heinrich committed:
feat: introduce personal settings import action

lukas-heinrich committed:
fix: add missing language keys

lukas-heinrich committed:
chore: create cleanup migration and adjust existing setup steps

lukas-heinrich committed:
refactor: adjust test settings usage outside of Test module

lukas-heinrich committed:
fix: denormalize array correctly

lukas-heinrich committed:
fix: import/export xml escaping

lukas-heinrich committed:
refactor: improve code style after review

Co-authored-by: Matheus Zych <mzych@databay.de>

lukas-heinrich committed:
fix: type error and missing export attributes

lukas-heinrich committed:
fix: rename creation column, fix actions for entire table and template import

lukas-heinrich committed:
fix: failing unit tests

lukas-heinrich committed:
fix: code style

dsstrassner commented:
Hi @lukas-heinrich, 

-  if I do not enter an author at the creation modal, I get the error message "Please enter a title for your template." not the error for a missing "author".
- I am pretty sure that this is not correct: 
<img width="623" height="316" alt="2025-10-20 08_58_33-General_ Settings_ Denis1_ ILIAS — Mozilla Firefox" src="https://github.com/user-attachments/assets/55039067-7790-4ecf-a457-55c950eb24e0" />

dsstrassner commented:
hi @lukas-heinrich & @thojou, maybe you missed it within the feature request, but this should be removed. There is no "Typ" dialog with ILIAS 11: 

<img width="1347" height="665" alt="2025-10-20 09_09_46-Test_ ILIAS — Mozilla Firefox" src="https://github.com/user-attachments/assets/5e7b6f0a-289b-40b7-8b30-1edc9ba456c4" />

dsstrassner commented:
- [x] Feature Request states "Table Header is changed to 'Test Setting Templates' instead 'Available test defaults'." Was this changed to "Personal Test Settings Templates" during the implementation? @oliversamoila 
- [x] Column order is not as stated in the feature request. Is this on purpose? @lukas-heinrich / @oliversamoila 
- [x] Naming of the action menu entries is not as in the FR? @lukas-heinrich / @oliversamoila

oliversamoila commented:
Thank you, @dsstrassner 
We have identified a small number of minor issues internally. I will clarify what is still open so that we are all on the same track.

Regarding Denis' feedback:
* @lukas-heinrich, please use: 
  * `Personal Test Settings Templates` instead of `Personal Default Settings` 
  * `Persönliche Testeinstellungsvorlagen` instead of `Persönliche Einstellungsvorlagen`
  * (This has been edited. Please pardon the confusion.)
* Please take a look at the column titles, the sorting options, and the labels for the actions. All of this is defined in the feature request under `Table Content` in the Conceptual Summary chapter. 
* Contrary to the description in the Feature Request, I ask you to also use `Personal Test Settings Templates` | `Persönliche Testeinstellungsvorlagen` for the table header. (This will ensure consistency.)

Thank all of you very much.

lukas-heinrich committed:
refactor: remove type dialog in test creation process, rename settings tab, table, columns and actions to enhance consistency

slowe47 commented:
Hi all of you,
approval by customer is done, see https://docu.ilias.de/go/wiki/wpage_8640_1357

dsstrassner commented:
@kergomard Please have a look!

kergomard reviewed (changes requested):
Thank you very much for the PR @lukas-heinrich !

This is a very nice improvement and do not be shocked by the long list of change requests. I'm sorry for them, but I think we can still improve a few things here, to not introduce issues we then regret. It was a long day of reading code, so let me know, if something is not understandable of what I wrote.

I think there is one biggy the rest is going to be small ones:
- [ ] I'm not comfortable with the `PropertyNormalizer`. Please get rid of it. This forces us to write constructors to satisfy extraneous requirements and we are back to some kind of `active_record`-pattern. I also think the `ContextualNormalizable` interface can go. I do not think we should talk about `Contracts` either, and finalize I'm not so sure about the `normalize`-nomenclature, as I do not think we are "normalizing" anything here (at least not in a mathematical sense). So: I would suggest to keep the Interface `Normalizable` but rename it `Exportable` with the functions `toExport()` and `fromExport()`. If you want to have a way to abstract the repeating tasks, I would suggest a `ExportValueConverter` that can be passed as parameter to both the functions. Put the `Interface` and the `ExportValueConverter` directly into `ExportImport`.

These are maybe also a slightly bigger thing:
- [ ] Please move  the migration of the old settings into the new settings from the DBUpdateSteps to the `RemoveLegacyTestSettings`-migration that then needs to be renamed. Make sure that test show a sensible error message, if the settings have not been migrated. The only thing that is a dbupdate step is the creation of the new table. We cannot do this in the UpdateSteps, it is explicitly not allowed and would make them run long.

Then to the smaller things:
- [ ] Could we please use something more verbose instead of "st" as alias for "tst_test_settings"? Maybe 'tst_set' or just no alias?
- [x] Please do not introduce a second postfix for local cmds. You can move all to . "Cmd", fi you want, but then please really move all.
- [ ] Please do not return '0' for a non-existing value in `getSelectedPersonalDefaultsSettingsFromForm()`. `null` is right in this case and we should not need to know that we are counting one based to know that 0 does mean invalid.
- [ ] Please do not call functions that build a modal "buildInput", but "buildModal". ...but I do not think we should be using modals in this case anyway, couldn't we implement this using the new `Prompts`?
- [ ] We should calling functions `execute()`(I know, I'm guilty of this, too. Please call the functions `perform()` (in other cases it might also be `run()`.
- [x] Please do not introduce the `TestDIC` in the middle of a class like e.g. [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10256/files#diff-ddf70ab522d8b2d87a5d9e8d13b382e4f4cedd91ca2bc23296fdaa2847117304R2113). It should only be used in the constructor. This here could also be solved by a little factory that  could be retrieved from the `TestDIC` in the constructor. Then the `->get()`-function of the `Factory` would only need a `URI` to return the whole shebang and we would get a lot of code out of `ilObjTestGUI`.
- [x] Please rename `MarkSchemaFactory::createSimpleSchema()` to `createSimplyeSchemaFromDBRow()` and move `MarkSchema->createSimpleSchema()` to this factory. And apply the same pattern to `SettingsFactory`.
- [ ] Could we please get rid of the `MainSettingsRepository::createFor()` again? This should be handled internally by `store` and should never need a second function from where I stand. Store should then simply return a clone of `MainSettings` with an id.
- [ ] Please set all properties to `private` unless you are absolutely sure, we would like somebody to derive from the corresponding class and to access the corresponding function.
- [ ] Please use '!== []' [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10256/files#diff-1ad2ae4eb4968f53ff8614b23ddc3057fc3dfefc4353ed6feff7deca4d8bba8aR167). Counting values in my eyes should be reserved for the cases where it is really needed.
- [ ] The function [`parseElementsRecursively`](https://github.com/ILIAS-eLearning/ILIAS/pull/10256/files#diff-1ad2ae4eb4968f53ff8614b23ddc3057fc3dfefc4353ed6feff7deca4d8bba8aR160-R190) does not remove any malicious input, but the xml could have been manipulated. So we need to be sure that there is no way that any of this ever gets shown without sanitation. Maybe some checks on strings wouldn't be such a bad idea?
- [ ] I do not think `PersonalSettingsRepository` should access any tables that are not `tst_test_defaults` directly.
- [ ] `getSettings()` and `getMarkSchema()` should not be available on the `PersonalSettingsRepository` and all functions should loose the `Template` in their name.
- [ ] `createTemplate` should be handled by `store` in `PersonalSettingsRepository`.
- [ ] Please remove `PersonalSettingsTable::records` and rename `sortRecords()` to `orderRecords()`. You will need to get the count from the Repository, but this you need anyway, as you want to know how many records there are in total and without any limits applied and you should limit and order the records on the database not here in the table.
- [ ] Please use the logged-in users date-format and do not build your own in `PersonalSettingsRepository`.
- [ ] The use of the `AdditionalInformationGenerator` in `PersonalSettingsTableShowAction` is dangerous as its use is not expected here. I would ask you to not do this. If you do this, then you need to update the Generator and to move it out of the logs, to make it understandable that it is not only about logging.
- [ ] Please remove the `CloneIntroduuctionAndClosingRemarksMIgration`. It is not needed anymore.

I hope I didn't miss too much. This was rather biggish ;-) .

Thanks again and best,
@kergomard

kergomard commented:
Just as an information: I've corrected the pattern for `Exportable` above, to get a more consistent dependency-structure.

lukas-heinrich committed:
refactor: improve code style

- remove second postfix for local cmds
- rename perform() to execute()
- remove usage of TestDIC in the middle of classes
- rename and move mark schema creation
- rename settings factory methods
- use private properties instead of protected
- remove CloneIntroductionAndClosingRemarksMMigration
- remove getSelectedPersonalDefaultsSettingsFromForm()

lukas-heinrich committed:
refactor: export/import process

- remove PropertyNormalizer trait
- remove ContextualNormalizable interface
- rename Normalizable to Exportable and rename methods
- sanitize xml input

lukas-heinrich committed:
refactor: restructure repository dependencies and move methods from personal settings to main/score settings and marks repository

lukas-heinrich committed:
refactor: split setup steps and create migrations to move test settings and legacy test defaults

lukas-heinrich commented:
Hi @kergomard ,

Thank you very much for the detailed review and feedback!

I have adjusted the places in the code, improved the repository responsibilities and introduced migrations instead of update steps. I have also adjusted export/import, removed the trait and renamed the interface. In addition, I added some string operations and XML options to improve the security.

Due to time constraints, I have not converted the modals to the prompt component but only renamed the methods as noted. In addition, I kept the date format in `PersonalSettingsRepository`, as this ensures that UTC is always used when writing/reading to the database. I think this is a good mechanism to keep the dates consistent.

PS: I will resolve the merge conflicts shortly and submit them via rebase and force push.

Best Regards
@lukas-heinrich

lukas-heinrich committed:
fix: code style after rebasing

kergomard commented:
Thank you very much for the update @lukas-heinrich !

I will give you partial feedbacks here, so that you can go on working on this, while I look at the rest. I think I need to go through the whole thing again:

- [ ] Could we please only have one migration. If not: Please explain why this is not possible.
~~- [ ] I'm currently fine with the try-catch-block when the Settings are missing, but please do not throw a `SettingsNotFound`-exception, but a `MigrationNotRun`-exception and show in the interface a message stating this. This makes it abundantly clear, that this check has to go again.~~ Sorry, was wrong on this. Please just correct the error message in the Exception "\nThis may be because the test settings have not yet been migrated.  Please try to run RemoveLegacyTestSettingsMigration if you have not already done so."

More as soon as I got through the whole thing.

Best,
 @kergomard

thojou commented:
> * Could we please only have one migration. If not: Please explain why this is not possible.

Hey @kergomard,

we decided to split migrations into 3 parts, because we wanted the transition process as transparent (and optional) as possible. One could just run migration for LegacyTestSettings and LegacyTestSettingsTemplates without ever running the RemoveLegacy... migration steps.

Also i think that auditing the process is a little bit easier, if we have 3 different, meaningfull migrations.
But, this is more by design rather then technical limitations.

Best regards,
@thojou

kergomard commented:
Thanks for the answer @thojou 

Ok, I can live with that.

Best,
@kergomard

kergomard reviewed (commented):
Thanks again @lukas-heinrich and @lukas-heinrich !

I went through this very quickly, just to have a look. I think there are a few things from the review that got lost:

>   Could we please use something more verbose instead of "st" as alias for "tst_test_settings"? Maybe 'tst_set' or just no alias?

>  We should stop calling functions `execute()` (I know, I'm guilty of this, too). Please call the functions `perform()` (in other cases it might also be `run()`).

> Could we please get rid of the MainSettingsRepository::createFor() again? This should be handled internally by store and should never need a second function from where I stand. Store should then simply return a clone of MainSettings with an id.

> The use of the AdditionalInformationGenerator in PersonalSettingsTableShowAction is dangerous as its use is not expected here. I would ask you to not do this. If you do this, then you need to update the Generator and to move it out of the logs, to make it understandable that it is not only about logging.

- [x] Sorry for my confusion about the date formats, I looked at too much code that day, but the place where the date format should be changed is [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10256/files#diff-306813cecc9e048e5616d344e59018bb9bfec9de5aca0e8fd7d03964e563ecd9R73) in `PersonalSettingsTable` you should get it from the current user, as it is for display. You are perfectly right about the repository, there it should always be UTC and we MUST ensure that.
- [x] Could you please remove all "normalization"-language from "Exportable". It is basically just a conversion to and from an array. There is nothing special about this array ;-) .
- [ ] I get an error when I try to run `setup.php update`: "General error: 1005 Can't create table `ilias_trunk`.`tst_defaults_marks` (errno: 121 "Duplicate key on write or update")" (there is an error in `ilDatabase` earlier due to a wrong namespace, I will see if there already is a PR and make one - @thojou already wrote one ;-) ).

Other then this the code looks ok to me. I will try to run a few more tests as soon as the error in the update is solved.

Best,
@kergomard

lukas-heinrich committed:
refactor: use users date format in table and clarify code documentation

lukas-heinrich commented:
Hi @kergomard ,

Thanks for the feedback. I have adjusted the method- and class comments in `SettingsNotFoundException` and `Exportable`. In addition, the table column will now display in the user's format.

Best,
@lukas-heinrich

kergomard commented:
Hi @lukas-heinrich 

Thank you very much! Please have a look at the other points above, too.

I was especially not able to run through the setup and I tried different things. I would be glad for a few pointers to allow me some tests.

Thank you very much and best,
@kergomard

lukas-heinrich committed:
refactor: rename action methods and fix 'Duplicate key' exception in Test11UpdateSteps

kergomard commented:
Good morning

I try to get a little further, here a quick recap of my experience:

-  If I ignore this error, remove the failed step from `il_db_steps` and run the update again. Things go through (which is obvious as now the table exists and the corresponding code is not run again), but then I run into two additional problems when trying to open tests WITHOUT running the migration:
- The first one is that tests on the dashboard are shown as linked, even if they are not available due to the missing migration.
- In the repository when opening a list view of an object containing a test I see: "Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '.obj_fi AS obj_fi FROM tst_test_settings INNER JOIN tst_tests ON tst_tests.se...' at line 4 QUERY: `SELECT score_reporting, reporting_date, show_grading_status, show_grading_mark, pass_deletion_allowed tst_tests.obj_fi AS obj_fi FROM tst_test_settings INNER JOIN tst_tests ON tst_tests.settings_id = tst_test_settings.id WHERE obj_fi IN (...)` (there is a coma missing in "/srv/http/ilias-trunk/components/ILIAS/Test/classes/class.ilObjTestAccess.php" line 404). ...and then all the fields in this table are ambiguous, as they still are in `tst_tests` and in `tst_test_settings`. Once these two things are fixed, lists work.
- On direct access to a test that is not migrated I get an error "Call to a member function getTitle() on null" in `ilObjTestGUI::addLocatorItems()`.
- Once I fix this I run into the same problem in the call to `ilObjectGUI::prepareOutput()` in "/srv/http/ilias-trunk/components/ILIAS/Test/classes/class.ilObjTestGUI.php" line 295.
- And once I comment out `prepareOutput` I run into a redirect and a "Response was already sent".

...this is when I then stopped testing.

Best,
@kergomard

lukas-heinrich committed:
fix: database exception in ScoreSettingsDatabaseRepository

lukas-heinrich commented:
Hi @kergomard,

Thank you very much for the additional feedback and thorough testing. I have incorporated the comments from above and hopefully also resolved the issues with the missing migration. I couldn't find a quick way to disable the link of a broken course on the dashboard. Instead, an error message is now displayed on the dashboard when attempting to open a course without settings. The same applies to tests within a course, group, etc. and when calling a test directly.

Regarding the `createFor` method in `MainSettingsRepository`, I'm uncertain how I can merge these methods. The `createFor` method differs from the `store` method in that there is no Settings-DTO yet that can be saved to the database. To achieve this, I would need to implement a new method in the factory to instantiate a new object for the test settings by default. This wasn't necessary before, since the default values from the database columns were taken as values. Additionally, the createFor method requires a Test-ID to create the relation between settings and tests. I consider this difficult to implement in the `store` method, as it is also used in contexts where there is no relation to a test (e.g., for templates). For these two reasons, I had decided to implement two methods.

I look forward to comments on this or hints if I'm overlooking something. Otherwise, thank you again for the thorough review!

Best
@lukas-heinrich

kergomard commented:
Thank you very much @lukas-heinrich 

Thanks for the update!

- [This here](https://github.com/ILIAS-eLearning/ILIAS/pull/10256/commits/da4de91ed964a7c551c321781192ef5590fceb52#diff-d6ec9058f0e568f45f7b556baef53000afcc92d56a37a554f26fd44e05dca885R283) doesn't look right, what am I missing?: Please just revert this commit. It just makes things worse. We would need more conceptual work to do this right and we do not have the time for this. The usage is very close to a legitimate usage. This will probably bite us later on. Could you please add a not to the places where you get data out of it and display it in the modal "le, 2025-10-27: This is not the right way to go, as a logging facility is used directly for presentation.". On a sidenote: You can expect me to hand you back a PR the instant something is called a "Utility". "Utility" usually translates to "Thing I don't know what to call and where to put." which in turn translates to "This probably shouldn't exist.".
- Sorry, we cannot do [this](https://github.com/ILIAS-eLearning/ILIAS/pull/10256/commits/009db64f3f3f4951a819612a82a692971ed0d917). Please revert the commit and find a more localized solution.
- Please absolutely rename the alias for the `tst_settings`-Table from "st" to something more verbose. I will not let this slide.
- Yes, I would like the factory to create a default settings object, the only thing it needs for this is to receive a `test_id` and then by calling the constructor for the settings with only the `test_id` it should receive a `Settings`-object with everything set to the defaults. The default values should not be defined by any defaults in the db, they should be defined by the defaults in the constructor(s), this way I get my information in the code and do not need to care much about the details in the database. If you look at the different classes in all Settings types, you will see that they follow this pattern.

Best,
@kergomard

kergomard commented:
Sorry, one more observation:
- In the case of cloning we should not save the default settings, but fill them first with the values, then create and save them in one go, ...but this probably is outside of the scope of this PR.

lukas-heinrich committed:
fix: note that AdditionalInformationGenerator as a logging facility should not be used for presentation

lukas-heinrich committed:
refactor: use more verbose aliases for tst_test_settings

lukas-heinrich committed:
refactor: remove MainSettingsRepository::createFor method

kergomard commented:
@lukas-heinrich : Please let me know, when I should look at this again as we need to get this done today.

lukas-heinrich committed:
fix: use error handling to catch missing migrations issues in ilObjTestGUI

lukas-heinrich commented:
Hi @kergomard ,

I think all the open points from the comments have now been implemented. I will test now in parallel whether I have overlooked any errors or created some new issues.

Best
@lukas-heinrich

kergomard commented:
Thank you very much @lukas-heinrich !

There is one missing point (I will start testing in parallel, too):
- Please revert the commit moving the `AdditionalInformationGenerator` to utilities.

Best,
@kergomard

kergomard commented:
So, we are getting there:

- [ ] There is an error in MoveTestSettingsMigration on line 53: It should say `self::SETTINGS_COLUMGS`
Additionally I get two sql errors:
- [ ] I get an error in `/srv/http/ilias-trunk/components/ILIAS/Tracking/classes/status/class.ilLPStatusTestPassed.php: "Column not found: 1054 Unknown column 'tst_tests.nr_of_tries' in 'SELECT' QUERY: SELECT tst_active.active_id, tst_active.tries, count(tst_sequence.active_fi) `sequences`, tst_active.last_finished_pass, CASE WHEN (tst_tests.nr_of_tries - 1) = tst_active.last_finished_pass THEN '1' ELSE '0' END is_last_pass FROM tst_active LEFT JOIN tst_sequence ON tst_sequence.active_fi = tst_active.active_id LEFT JOIN tst_tests ON tst_tests.test_id = tst_active.test_fi WHERE tst_active.user_fi = 314 AND tst_active.test_fi = 473 GROUP BY tst_active.active_id, tst_active.tries, is_last_pass" and
- [ ] In `/srv/http/ilias-trunk/components/ILIAS/Test/src/Participants/ParticipantRepository.php` on line 227: "Integrity constraint violation: 1052 Column 'active_id' in WHERE is ambiguous QUERY: SELECT participants.*, manscoring_done.done as scoring_finalized FROM ( ( SELECT ta.active_id, ta.user_fi, ta.test_fi, ta.anonymous_id, ta.tries, ta.submitted, ta.last_finished_pass, ta.last_started_pass, COALESCE(ta.last_started_pass, -1) <> COALESCE(ta.last_finished_pass, -1) as unfinished_attempts, ud.firstname, ud.lastname, ud.login, ud.matriculation, (SELECT MIN(started) FROM tst_times WHERE active_fi = ta.active_id) as first_access, (SELECT MAX(finished) FROM tst_times WHERE active_fi = ta.active_id) as last_access, tatime.additionaltime extra_time, tinvited.ip_range_from, tinvited.ip_range_to, tinvited.tstamp as invitation_date FROM tst_active ta LEFT JOIN usr_data ud ON ud.usr_id = ta.user_fi LEFT JOIN tst_addtime tatime ON tatime.user_fi = ta.user_fi AND tatime.test_fi = ta.test_fi LEFT JOIN tst_invited_user tinvited ON tinvited.test_fi = ta.test_fi AND tinvited.user_fi = ta.user_fi WHERE ta.test_fi = 386 ) UNION ( SELECT ta.active_id, tinvited.user_fi, tinvited.test_fi, ta.anonymous_id, ta.tries, ta.submitted, ta.last_finished_pass, ta.last_started_pass, COALESCE(ta.last_started_pass, -1) <> COALESCE(ta.last_finished_pass, -1) as unfinished_attempts, ud.firstname, ud.lastname, ud.login, ud.matriculation, NULL as first_access, NULL as last_access, tatime.additionaltime extra_time, tinvited.ip_range_from, tinvited.ip_range_to, tinvited.tstamp as invitation_date FROM tst_invited_user tinvited LEFT JOIN usr_data ud ON ud.usr_id = tinvited.user_fi LEFT JOIN tst_addtime tatime ON tatime.user_fi = tinvited.user_fi AND tatime.test_fi = tinvited.test_fi LEFT JOIN tst_active ta ON tinvited.test_fi = ta.test_fi AND tinvited.user_fi = ta.user_fi WHERE tinvited.test_fi = 386 AND ta.active_id IS NULL ) ) as participants LEFT JOIN manscoring_done ON manscoring_done.active_id = participants.active_id WHERE active_id = 192"

I will then merge and we will deal with the rest during testing.

Thanks and best,
@kergomard

lukas-heinrich committed:
fix: wrong identifier in MoveTestSettingsMigration and invalid sql queries

lukas-heinrich commented:
Hi @kergomard ,

Thank you for the feedback. I have adjusted the three places and fixed the SQL errors. 
The commit with the AdditionalInformationGenerator changes should no longer be in the Git history, and I replaced it with d2cbad3423df12fcc1d8187b7b3903331af04d8d.

Many thanks for your efforts and review!

Best,
@lukas-heinrich

kergomard commented:
Thank you very much @lukas-heinrich !

Sorry that I missed the removal of the `AdditionalInformationGenerator` commit! I was so fixated on a revert commit that I didn't look anywhere else!

I will squash this here as the history is on shaky grounds.

Thanks again for all your patience and best,
@kergomard
